### PR TITLE
Move EmbeddedChannel to its own library.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

EmbeddedChannel is an important testing tool, and we want to use it
without needing to bring along the POSIX layer. They are not tightly
coupled. However, it also doesn't belong naturally in NIOCore, so we
should probably put it in its own place.

Modifications:

- Moved EmbeddedChannel and EmbeddedEventLoop to NIOEmbedded.
- Moved the tests to NIOEmbeddedTests
- Duplicated some test helpers

Result:

Easy to use EmbeddedChannel without the POSIX layer.

### CODE REVIEW

File: Tests/NIOEmbeddedTests/EmbeddedChannelTest.swift [hunk 1]
```diff
@@ -13,7 +13,60 @@
 //===----------------------------------------------------------------------===//
 
 import XCTest
-@testable import NIO
+import NIOCore
+@testable import NIOEmbedded
+
+class ChannelLifecycleHandler: ChannelInboundHandler {
```
Comment (fabianfett) [inline Tests/NIOEmbeddedTests/EmbeddedChannelTest.swift:19]: Is that used anywhere?

Comment (Lukasa) [reply Tests/NIOEmbeddedTests/EmbeddedChannelTest.swift:19]: Yup, in the tests within this file.

File: Package.swift [hunk 2]
```diff
@@ -19,13 +19,15 @@ var targets: [PackageDescription.Target] = [
     .target(name: "NIOCore",
             dependencies: ["NIOConcurrencyHelpers", "CNIOLinux"]),
     .target(name: "_NIODataStructures"),
+    .target(name: "NIOEmbedded", dependencies: ["NIOCore", "_NIODataStructures"]),
```
Comment (PeterAdams-A) [inline Package.swift:22]: Don't like the name.

This is all about Testing really?  Embedded also don't give much scope for adding things here.  Also makes me think of embedded systems.

File: Tests/NIOEmbeddedTests/EmbeddedChannelTest.swift (hunk 1 above)
Comment (fabianfett) [reply Tests/NIOEmbeddedTests/EmbeddedChannelTest.swift:19]: Found it! Thanks! Where did this code live before? Is this now duplicated?

Comment (Lukasa) [reply Tests/NIOEmbeddedTests/EmbeddedChannelTest.swift:19]: Yup, it lived in NIO’s channel tests. I don’t think we’re likely to bother deduplicating it though, it’s fairly trivial.

File: Package.swift (hunk 2 above)
Comment (Lukasa) [reply Package.swift:22]: That’s true, but I don’t want us to push too hard on testing: while this is its primary use, there are reasons you might want to use it in production code.

Maybe `NIOSingleThreaded`?

Comment (glbrntt) [reply Package.swift:22]: I don't mind the `NIOEmbedded` name.

I wonder if explaining the various "main" NIO modules (`NIO`, `NIOCore`, `NIOEmbedded`, `NIOFoundationCompat` etc) in the README would be useful addition.

**Review by glbrntt — APPROVED** [review]

File: Package.swift (hunk 2 above)
Comment (PeterAdams-A) [reply Package.swift:22]: What's the real reason why it doesn't fit anywhere else?  Is regular EL still in NIO[Posix]?

Comment (Lukasa) [reply Package.swift:22]: Define "regular EL"?

Comment (Lukasa) [reply Package.swift:22]: The `EventLoop` protocol is in NIOCore. The SelectableEventLoop is in NIO, and will stay there as it's inherently POSIX-specific.

Comment (Lukasa) [reply Package.swift:22]: Discussed in a stand-up meeting, agreed to go ahead.

**Pull request merged**